### PR TITLE
#comment add custom authorizable to notifications

### PR DESCRIPTION
**Pull request by noodlemoodle**
foam pr for added notification permissions https://github.com/nanoPayinc/NANOPAY/pull/7521

### CODE REVIEW

**Review by jlhughes — APPROVED** [review]
Perfect.

File: src/foam/nanos/notification/Notification.js [hunk 1]
```diff
@@ -114,5 +120,52 @@ foam.CLASS({
...
+      ],
+      type: 'Boolean',
+      javaCode: `
+        foam.nanos.auth.User user = (foam.nanos.auth.User) x.get("user");
+        if( user == null ) return false;
+
+        Long userId = user.getId();
+        String groupId = user.getGroup();
+
+        return getUserId() != userId && getGroupId() != groupId;
```
Comment (smirnovaka) [inline src/foam/nanos/notification/Notification.js]: `grroupId` property is only used as a helper to define what users should receive notifications but it is reset to null prior being saved to journals. You should only check for userId, this line will always return false

File: src/foam/nanos/notification/Notification.js [hunk 2]
```diff
@@ -114,5 +120,52 @@ foam.CLASS({
...
+      `
+    },
+    {
+      name: 'authorizeOnCreate',
+      javaCode: `
+      AuthService auth = (AuthService) x.get("auth");
+      if ( ! auth.check(x, "*") ) throw new AuthorizationException("You don't have permission to create this notification.");
```
Comment (smirnovaka) [inline src/foam/nanos/notification/Notification.js]: please create permissions for create/read rather than *

**Review by smirnovaka — CHANGES_REQUESTED** [review]

**Comment by smirnovaka** [comment]
Great job @noodlemoodle, requested some changes. Also still thinking about making it more generic and use it for other services too. @jlhughes?

File: src/foam/nanos/notification/Notification.js (hunk 2 above)
Comment (noodlemoodle) [reply src/foam/nanos/notification/Notification.js]: added permissions https://github.com/nanoPayinc/NANOPAY/pull/7506

File: src/foam/nanos/notification/Notification.js [hunk 3]
```diff
@@ -114,5 +120,49 @@ foam.CLASS({
...
+      javaCode: `
+      AuthService auth = (AuthService) x.get("auth");
+      if ( ! auth.check(x, "*") ) throw new AuthorizationException("You don't have permission to delete notifications you do not own.");
+      `
+    },
+    {
+      name: 'authorizeOnRead',
+      javaCode: `
+      AuthService auth = (AuthService) x.get("auth");
+      if ( ! checkOwnership(x) && ! auth.check(x, "notificationDAO.read.*") ) throw new AuthorizationException("You don't have permission to read notifications you do not own.");
```
Comment (jlhughes) [inline src/foam/nanos/notification/Notification.js]: @penzital @noodlemoodle Is authorizeOnRead called on every object put in the sink on select()?  Curious how checkOwnership works?

Comment (noodlemoodle) [reply src/foam/nanos/notification/Notification.js]: Yes, the authorizer containing this is passed into the authorizationDAO which adds HAS_PERMISSION as part of the predicate in select. HAS_PERMISSION calls the authorizeOnRead on every object before putting it into the sink I believe. this predicate was added in this pr https://github.com/foam-framework/foam2/pull/2395
so in this case a read on a notification would be granted if the notification belong to current user or current user has admin permission.

Comment (smirnovaka) [reply src/foam/nanos/notification/Notification.js]: authorizeOnRead is called on every object unless current user has global read permission

Comment (smirnovaka) [inline src/foam/nanos/notification/Notification.js]: you don't need to check for global read permission here, `AuthorizableAuthorizer.checkGlobal` will do it for you. This method won't be even called if user has `notification.read.*` in these methods you actually need to check for `read.id`

**Review by smirnovaka — CHANGES_REQUESTED** [review]
this pr will be refactored anyway

File: src/foam/nanos/notification/Notification.js (hunk 3 above)
Comment (noodlemoodle) [reply src/foam/nanos/notification/Notification.js]: oh true true

**Pull request merged**